libbeat/common/event.go

```diff
@@ -1,92 +1,223 @@
+// that are found in the MapStr are dropped and warnings are logged.
+func ConvertToGenericEvent(m MapStr) MapStr {
+	event, errs := normalizeMap("", m)
+	if len(errs) > 0 {
+		logp.Warn("Unsuccessful conversion to generic event: %v errors: %v, "+
+			"event=%#v", len(errs), errs, m)
 	}
-	var v1 MapStr
-	err = json.Unmarshal(marshaled, &v1)
-	if err != nil {
-		logp.Warn("unmarshal err: %v", err)
-		return nil, err
+	return event
+}
+
+// normalizeMap normalizes each element contained in the given map. If an error
+// occurs during normalization, processing of m will continue, and all errors
+// are returned at the end.
+func normalizeMap(baseKey string, m MapStr) (MapStr, []error) {
+	var errs []error
+
+	out := make(MapStr, len(m))
+	for key, value := range m {
+		fullKey := joinKeys(baseKey, key)
+		v, err := normalizeValue(fullKey, value)
+		if len(err) > 0 {
+			errs = append(errs, err...)
+		}
+
+		// Drop nil values from maps.
+		if v == nil {
+			if logp.IsDebug(eventDebugSelector) {
+				eventDebugf("Dropped nil value from event where key=%v", fullKey)
+			}
+			continue
+		}
+
+		out[key] = v
 	}
 
-	return v1, nil
+	return out, errs
 }
 
-func ConvertToGenericEvent(v MapStr) MapStr {
+// normalizeMapStrSlice normalizes each individual MapStr.
+func normalizeMapStrSlice(baseKey string, maps []MapStr) ([]MapStr, []error) {
+	var errs []error
 
-	for key, value := range v {
+	out := make([]MapStr, 0, len(maps))
+	for i, m := range maps {
+		normalizedMap, err := normalizeMap(joinKeys(baseKey, strconv.Itoa(i)), m)
+		if len(err) > 0 {
+			errs = append(errs, err...)
+		}
+		out = append(out, normalizedMap)
+	}
 
-		if value == nil {
-			// leave nil values alone
-			continue
+	return out, errs
+}
+
+// normalizeMapStringSlice normalizes each individual map[string]interface{} and
+// returns a []MapStr.
+func normalizeMapStringSlice(baseKey string, maps []map[string]interface{}) ([]MapStr, []error) {
+	var errs []error
+
+	out := make([]MapStr, 0, len(maps))
+	for i, m := range maps {
+		normalizedMap, err := normalizeMap(joinKeys(baseKey, strconv.Itoa(i)), m)
+		if len(err) > 0 {
+			errs = append(errs, err...)
 		}
+		out = append(out, normalizedMap)
+	}
 
-		switch value.(type) {
-		case Time, *Time:
-			continue
-		case time.Location, *time.Location:
-			continue
-		case MapStr:
-			v[key] = ConvertToGenericEvent(value.(MapStr))
-			continue
-		case *MapStr:
-			v[key] = ConvertToGenericEvent(*value.(*MapStr))
-			continue
-		default:
+	return out, errs
+}
 
-			typ := reflect.TypeOf(value)
+// normalizeSlice normalizes each element of the slice and returns a []interface{}.
+func normalizeSlice(baseKey string, v reflect.Value) (interface{}, []error) {
+	var errs []error
+	var sliceValues []interface{}
 
-			if typ.Kind() == reflect.Ptr {
-				typ = typ.Elem()
-			}
+	n := v.Len()
+	for i := 0; i < n; i++ {
+		sliceValue, err := normalizeValue(joinKeys(baseKey, strconv.Itoa(i)), v.Index(i).Interface())
```

Comment:
Good question, I was think we might want to leave arrays alone in case the index conveys meaning. For example, if some application using structured logging writes an array of values from various sensors, dropping one of those values could cause problems.